File: Python/ExampleTests/common/new_post_string.py
from . import tokens
DNC_STRING = '?DNC?'


class NewPostString:
    def __init__(self, title: str = '', body: str = '', user_id: str = ''):
        self.title = title
        self.body = body
        self.user_id = user_id

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 3, 'NewPost')
        return cls(
            parts[0],
            parts[1],
            parts[2]
        )

    def __str__(self):
        return (tokens.token(self.title) + ' ' +
                tokens.token(self.body) + ' ' +
                tokens.token(self.user_id))

    def _key(self):
        return (self.title, self.body, self.user_id)

    def __eq__(self, other):
        if not isinstance(other, NewPostString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'NewPostString({self})'

File: Swift/ExampleTests/common/NewPostString.swift
public struct NewPostString: CustomStringConvertible, Equatable {
    public let title: String
    public let body: String
    public let userId: String

    public init(title: String, body: String, userId: String) {
        self.title = title
        self.body = body
        self.userId = userId
    }

    public init(fromArray v: [String]) {
        self.title = v.count > 0 ? v[0] : ""
        self.body = v.count > 1 ? v[1] : ""
        self.userId = v.count > 2 ? v[2] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> NewPostString {
        let parts = Tokens.require(text, 3, "NewPost")
        return NewPostString(title: parts[0], body: parts[1], userId: parts[2])
    }

    public var description: String {
        return Tokens.token(title) + " " + Tokens.token(body) + " " + Tokens.token(userId)
    }

    public static let dncString = "?DNC?"

    public static func == (a: NewPostString, b: NewPostString) -> Bool {
        return (a.title == dncString || b.title == dncString || a.title == b.title)
            && (a.body == dncString || b.body == dncString || a.body == b.body)
            && (a.userId == dncString || b.userId == dncString || a.userId == b.userId)
    }
}
